### PR TITLE
merge Dan's PR of determinize

### PR DESCRIPTION
**Pull request by qindazhu**
Note: this PR is totally from @danpovey https://github.com/danpovey/k2/pull/37, I'm just fixing some compiler errors.

Nearly fixed with 3 group errors left. Please see comments below.

### CODE REVIEW

File: k2/csrc/determinize.cc [hunk 1]
```diff
@@ -5,469 +5,287 @@
...
+    std::vector<std::vector<typename TracebackState::DerivType>>
+        *derivs_per_arc,
+    DetStateMap<TracebackState> *state_map) {
+  float arc_weight;
+  std::vector<typename TracebackState::DerivType> deriv_info;
+  state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+  int32_t next_state_id;
+  bool is_new_state = state_map->GetOutputState(state);
+  arcs_out->push_back({this->state_id, next_state_id, label});
+  arc_weights_out->push_back(arc_weight);
```
Comment (qindazhu) [inline k2/csrc/determinize.cc:137]: What does `this` point to? and what's the `next_state_id` , seems we need to change the signature of `GetOutputState` to return a pair of `<bool, next_state_id>`.

And what's the `label`

File: k2/csrc/determinize.cc [hunk 2]
```diff
@@ -5,469 +5,287 @@
...
+       ++iter) {
+    DetState<TracebackState> *det_state = iter->second;
+
+    float arc_weight;
+    std::vector<DerivType> deriv_info;
+    det_state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+    if (det_state->forward_backward_prob >= prune_cutoff) {
+      bool is_new_state = state_map->GetOutputState(det_state);
+      arcs_out->push_back({this->state_id, next_state_id, label});
+      arc_weights_out->push_back(arc_weight);
```
Comment (qindazhu) [inline k2/csrc/determinize.cc:193]: again, `next_state_id`, `label` and `arc_weight`

File: k2/csrc/determinize.cc [hunk 3]
```diff
@@ -5,469 +5,287 @@
...
+  using DS = DetState<TracebackState>;
+
+  std::shared_ptr<DS> start_state = std::make_shared<DS>();
+
+  std::vector<Arc> arcs_out;
+  arc_weights_out->clear();
+  arc_derivs_out->clear();
+
+  bool ans = map.GetOutputState(start_state.get());
+  CHECK(ans && ans->state_id == 0);
```
Comment (qindazhu) [inline k2/csrc/determinize.cc:268]: again, seems we should change the signature of `GetOutputState`?

Comment (danpovey) [reply k2/csrc/determinize.cc:268]: Yes I think so (add wfsa_in.fsa)

File: k2/csrc/determinize.h [hunk 4]
```diff
@@ -0,0 +1,686 @@
...
+             b = d.base_state * 103979 + d.seq_len;
+
+    // We choose an arbitrary DetStateElement (the first one in the list) to
+    // read the symbol sequence from; the symbol sequence will be the same no
+    // matter which element we choose to trace back.
+    auto elem = d.elements.begin()->second;
+    int32_t seq_len = d.seq_len;
+    const auto &arcs = fsa.arcs;
+    for (int32_t i = 0; i < seq_len; ++i) {
+      int32_t symbol = arcs[elem->arc_id].label;
```
Comment (qindazhu) [inline k2/csrc/determinize.h:634]: as `MaxTracebackState` and `LogSumTracebackState` has differnt structure, we may need to define differnt functions for them, For `LogSum`, how do compute the hash pair as there are a vector of prev_elements `vector<LogSumTracebackLink> prev_elements`, iterover all elements in it?

Comment (danpovey) [reply k2/csrc/determinize.h:634]: No you can just pick an arbitrary one, the symbol-sequence will be the same no matter which you pick.

Comment (qindazhu) [reply k2/csrc/determinize.h:753]: OK, I just picked the first one.

File: k2/csrc/determinize.h [hunk 5]
```diff
@@ -0,0 +1,837 @@
...
+    auto ret = elements.insert({state_id, nullptr});
+    if (!ret.second) {  // No such state existed in `elements`
+      ret.first->second = std::make_shared<TracebackState>(
+          state_id, src, incoming_arc_index, arc_weight);
+    } else {  // A state with this staste_id existed in `elements`.
+      ret.first->second->Accept(src, incoming_arc_index, arc_weight);
+    }
+  }
+
+  int32_t state_id;
```
Comment (qindazhu) [inline k2/csrc/determinize.h:459]: @danpovey , I added a property of `state_id` as it seems it is used in many places (but do definition)

File: k2/csrc/determinize.h [hunk 6]
```diff
@@ -0,0 +1,837 @@
...
+             b = d.state_id * 103979 + d.seq_len;
+
+    // We choose an arbitrary DetStateElement (the first one in the list) to
+    // read the symbol sequence from; the symbol sequence will be the same no
+    // matter which element we choose to trace back.
+    auto elem = d.elements.begin()->second;
+    int32_t seq_len = d.seq_len;
+    const auto &arcs = fsa.arcs;
+    for (int32_t i = 0; i < seq_len; ++i) {
+      int32_t symbol = arcs[elem->prev_elements[0].arc_index].label;
```
Comment (qindazhu) [inline k2/csrc/determinize.h:776]: just pick the first element to get the hash

**Comment by qindazhu** [comment]
OK, fixed all compiler errors and style issues, ready to merge (so there will no style issues for following PR from other guys). Definitely there will be bugs (and memory leaks as I noticed there are usages of `new` without `unique_ptr`). We'll need to write some tests to debug left issues.

File: k2/csrc/determinize.h (hunk 5 above)
Comment (danpovey) [reply k2/csrc/determinize.h:459]: OK but it needs to be documented.  E.g.
```
  // State-id in the output FSA.  Only defined after DetStateMap::GetOutputState()
  //  is called.  A DetState that is in the queue should have this defined.
```

Comment (qindazhu) [reply k2/csrc/determinize.h:462]: Added

**Comment by danpovey** [comment]
Great.  Would you mind making sure the code runs?  E.g. making some test cases?

**Comment by qindazhu** [comment]
I think we can merge it first to make other PRs pass the style check. I'll make another PR to debug with test cases this weekend, is that OK?

**Comment by danpovey** [comment]
Sure, thanks!

**Pull request merged**